### ISSUE TITLE
Typedef Struct

### ISSUE THREAD

**Issue by kaendfinger** [mezoni/binary_declarations#2]
Hello again! I'm sure you are already aware but I am going to use your libraries to make a libgit2 binding in pure Dart. I'm using your binary declarations but I noticed you don't support typedef structs. Will you be adding support for this?

Here is what I have:

```c
typedef struct {
  char *ptr;
  size_t asize;
  size_t size;
} git_buf;

void git_buf_free(git_buf *buffer);
int git_buf_set(git_buf *buffer, const void *data, size_t datalen);
int git_config_find_global(git_buf *out);
```

**Comment by mezoni** [comment]
Already added in version 0.0.4

```dart
import "package:binary_declarations/binary_declarations.dart";

void main() {
  var declarations = new BinaryDeclarations(text);
  for (var declaration in declarations) {
    print(declaration);
  }
}

var text = '''
signed char[10] foo(struct S);

struct _s {} s;
int i;
int* ip;
void* vpa;
struct _s { int i; } s;
typedef int INT;
typedef void* PVOID;
typedef void* PVOIDA[];
typedef void* PVOIDA[10];
typedef struct _s SA[];
typedef struct {} S;
typedef struct _s {} S2;
typedef struct _s { int i; int* ip; } S2;
typedef struct _s { int i; int* ip; struct s { int i; } s; } S2;
struct {};
struct _s {};
struct _s {
  int i;
  char* cp;
};
int foo(void);
int foo(void* []);
int foo(char[]);
int foo(int, ...);
int foo(int, const char*, ...);
int foo(const int);
int foo(int i);
int foo(int*, char[]);
int foo(int* ip, char cp[]);
int foo(int* ip, char cp[10]);
int foo(int* ip, char cp[10], struct S);
int foo(int* ip, char cp[10], union S);
struct S foo(int* ip, char cp[10][3], struct S);
struct S[] foo(int* ip, char cp[10][3], struct S);
signed char[10] foo(int* ip, char cp[10][3], struct S);
''';
```

Output

```
signed char[10] foo(struct S)
struct _s { } s
int i
int* ip
void* vpa
struct _s {  int i; } s
typedef int INT
typedef void* PVOID
typedef void* PVOIDA[]
typedef void* PVOIDA[10]
typedef struct _s SA[]
typedef struct { } S
typedef struct _s { } S2
typedef struct _s {  int i; int* ip; } S2
typedef struct _s {  int i; int* ip; struct s {  int i; } s; } S2
struct { }
struct _s { }
struct _s {  int i; char* cp; }
int foo()
int foo(void*[])
int foo(char[])
int foo(int, ...)
int foo(int, char*, ...)
int foo(const int)
int foo(int i)
int foo(int*, char[])
int foo(int* ip, char cp[])
int foo(int* ip, char cp[10])
int foo(int* ip, char cp[10], struct S)
int foo(int* ip, char cp[10], union S)
struct S foo(int* ip, char cp[10][3], struct S)
struct S[] foo(int* ip, char cp[10][3], struct S)
signed char[10] foo(int* ip, char cp[10][3], struct S)
```

**Comment by mezoni** [comment]
Possible this feature still not added into "binary_types".
Tomorow I will fix this.

**Comment by kaendfinger** [comment]
Ok :)

**Comment by mezoni** [comment]
Added initial support.

```dart
var _header = '''
struct struct_t {
  struct { int i; } s;          
  int8_t int8;
  int16_t int16;
  int32_t int32;
  int64_t int64;
  uint8_t uint8;
  uint16_t uint16;
  uint32_t uint32;
  uint64_t uint64;
  int* ip;
  void* vp;           
};

// Should be also declared
struct _Point {
  int x;
  int y;
} unusedVar;

typedef struct _Point POINT;

typedef struct {
  int i;
} struct1_t;

typedef struct {
  int i;
} struct2_t;

typedef struct {
  int i;
} struct4_t;

typedef struct {
  POINT a;
  POINT b;
} RECT;

typedef struct { int i;
  int* ip;
  int ia[10];
  POINT pt;
} struct6_t;

struct struct8_t {
  struct struct8_t *s;
};
''';
```

**Comment by kaendfinger** [comment]
Thanks

**Comment by mezoni** [comment]
You can practice a bit with multi-level access to data.

Over the next few days I will rework (rollback a work of last months) "binary types" to make them more suitable for using as a data for a higher-level data access objects such as binary classes (not a name of yet another "binary" package) but a some "yet  another kind of the data accessors".

Of course, these classes should be written by the hands or by a some tools (hint: my "binary generator" but not limited only on it).

Initially "binary types" was ready for that but but I didn't thought at the time about a such feature as a "wrapping". I rework them (not so long ago) for better an easier access data through them but this made them not very suitable for wrapping.

Wrapped (into classes) data does not relieve you from having to think about the low level binary types, but you'll be less likely to remember them.
And, of course, they will work directly with a physical memory. Not so in the case of .NET pinvoke but I also think about such level of programming. At least about how easy convert binary data (even a wrapped) into "ready to use for end-user". But not at this time.

And, of course, they will work directly with a physical memory. Not so in the case of .NET pinvoke but I also think about such level of programming. At least about how easy convert binary data (even a wrapped) into "ready to use for end-user" (regular classes).
We can can call them as  a "binary converters" on the fly.
This will be third layer in data accessing (end-user presentation layer).
But, of course, not at this time.

Wait a few days and can try initial releases of binary generator.

P.S.

**You can already try my links to the donation.**
Symbolically, $1.

Another hint. Binary support.
Simplify the search distributed binary files (dynamic libraries) in the "pub cache" depending on the current architecture and version of Dart package.

You can distribute them in separate packages. Why need include them (many megabytes) each time in pure Dart packages?

**Comment by mezoni** [comment]
About breaking change (roll back).

Binary types is not a copy of C types.
They based on such idea but they differs.
Currently they closer to C types.
But if we want wrap them we need "objective binary types.
This is because we need some control on a memory bounds access.
Arrays in binary types is not the same as arrays in C types.
They compatible but in C you can omit "&" on the array.
Currently this work but I will have revert this back where array need to be directly referenced.

Eg.

This currently work

```
int *ip;
int ia[1];
ip = ia
ip = ip
```

In new vserion (rolled back ) should requires direct "&"

```
int *ip;
int ia[1];
ip = &ia // <========== &
ip = ip
```

P.S.

In short words, I will have
- Remove "Reference" class in favor to "Binary Data".
- Rework "Array Type" class in favor to object-oriented array.
Object-oriented array (array of array of any data) means that in physical memory everything will be stored the same as with C language.
But on the Dart side this will be a slightly better organized data but at the same time fully compatible with C data.
This means that binary types is subset of "C types" (C extension) except on how to operate with an array when you want pass it by reference (&ia).
